### PR TITLE
Taskwait nowait Test

### PR DESCRIPTION
**Pull request by mjcarr458**
…ze nowait clause

### CODE REVIEW

File: tests/5.1/taskwait/test_task_nowait.c [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include "ompvv.h"
+#include <math.h>
+
+#define N 1024
+
+int test_task_nowait(){
+	int errors = 0;
+	int test_scaler = 1;
```
Comment (tmh97) [inline tests/5.1/taskwait/test_task_nowait.c:21]: scalar*

File: tests/5.1/taskwait/test_task_nowait.c [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			}
+		}
+	}
+	int new_sum;
+	for (int i = 0; i < N; i++){
+		new_sum += test_arr[i];
+	}
+	OMPVV_TEST_AND_SET(errors, test_scaler != 2);
+	OMPVV_INFOMSG_IF(test_scaler == 1, "Scalar taskwait region was not ran");
+	OMPVV_INFOMSG_IF(test_scaler > 2, "Scalar taskwait region was ran more than once");
```
Comment (tmh97) [inline tests/5.1/taskwait/test_task_nowait.c:49]: if == 2 it was ran only once, add a separate check for that.

File: tests/5.1/taskwait/test_task_nowait.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+			for (int i=0; i<N; i++){
+				test_arr[i] = i;
+			}
+		}
+	}
+	int new_sum;
+	for (int i = 0; i < N; i++){
+		new_sum += test_arr[i];
+	}
+	OMPVV_TEST_AND_SET(errors, test_scaler != 2);
```
Comment (tmh97) [inline tests/5.1/taskwait/test_task_nowait.c:47]: OMPVV_TEST_AND_SET_VERBOSE

File: tests/5.1/taskwait/test_task_nowait.c [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	
+	#pragma omp parallel
+	{
+		#pragma omp taskwait nowait
+		{
+			test_scaler += 1;
+		}
+
+		#pragma omp taskwait nowait
```
Comment (tmh97) [inline tests/5.1/taskwait/test_task_nowait.c:36]: Change to #pragma omp task and then use taskwait in-between!

**Comment by mjcarr458** [comment]
add bottom of header and nowait clause

File: tests/5.1/taskwait/test_task_nowait.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
+//--------------- test_taskwait_nowait.c---------------------------------//
+//
+// OpenMP API Version 5.1 Aug 2021
+//
+// Tests the behavior of taskwait when nowait is specified.
+// The behavior of this should be similar to that of using task construct.
+// It is important to note that all depenend tasks must be finished first.
+
```
Comment (tmh97) [inline tests/5.1/taskwait/test_task_nowait.c:8]: add bottom of the header :)

File: tests/5.1/taskwait/test_task_nowait.c [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	int sum = 0;
+	for (int i =0; i<N; i++){
+		test_arr[i] = 1;
+		sum += i;
+	}
+	#pragma omp task private(test_scaler)
+	{
+		test_scaler += 1;
+	}
+	#pragma omp taskwait
```
Comment (tmh97) [inline tests/5.1/taskwait/test_task_nowait.c:32]: no wait!

**Comment by tob2** [comment]
By the way, in my understanding – and also when looking at the OpenMP Specification issue (issue 1771, not publicly available) – the main purpose of `nowait` is to use it with `omp taskwait depend(...)`.

The testcase uses a plain `omp taskwait nowait` — that's useful as parse test and an implementation should also support this. However, I think _adding_ a test which uses it _together with_ a `depend` clause makes sense – as this is the primary real-world use.

**Comment by jakubjelinek** [comment]
`#pragma omp taskwait nowait`
without depend clauses is invalid, there is a restriction:
"The nowait clause may only appear on a taskwait directive if the depend clause is present."
in both OpenMP 5.1 and 5.2.

File: tests/5.1/taskwait/test_task_nowait.c [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+		test_scaler += 1;
+	}
+	#pragma omp task depend(in : test_arr) shared(test_arr)
+	{
+		for (int i=0; i<N; i++){
+			test_arr[i] += 1;
+		}
+	}
+	#pragma omp taskwait nowait depend(in : test_scaler, test_arr)
+	int new_sum;
```
Comment (tob2) [inline tests/5.1/taskwait/test_task_nowait.c:40]: Ensure the variable is 0 at the beginning:
```suggestion
	int new_sum = 0;
```

File: tests/5.1/taskwait/test_task_nowait.c [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	int test_arr[N];
+	int sum = 0;
+	for (int i =0; i<N; i++){
+		test_arr[i] = 1;
+		sum += i;
+	}
+	#pragma omp parallel
+	#pragma omp single
+	{
+		#pragma omp task depend(out: test_scaler) shared(test_scaler)
```
Comment (spophale) [inline tests/5.1/taskwait/test_task_nowait.c:33]: the dependence should be inout as it is modifying an existing value.

Comment (jakubjelinek) [reply tests/5.1/taskwait/test_task_nowait.c:33]: inout and out are exactly the same under the hood / in the standard, so it is just an style issue.

**Review by spophale — APPROVED** [review]

**Review by tmh97 — APPROVED** [review]

**Pull request merged**